File: Python/dal/reader_category_service.py
from dal.db_helper import DBHelper
from models.reader_category import ReaderCategory


class ReaderCategoryService:
    def __init__(self):
        self.db = DBHelper()

    def get_all_categories(self) -> list:
        sql = "SELECT CategoryId, CategoryName, MaxBorrowCount, BorrowDays, LateFeePerDay FROM ReaderCategory ORDER BY CategoryId"
        rows, _ = self.db.execute_data_set(sql)
        return rows

    def add_category(self, category: ReaderCategory) -> int:
        sql = """
            INSERT INTO ReaderCategory (CategoryName, MaxBorrowCount, BorrowDays, LateFeePerDay)
            VALUES (?, ?, ?, ?)
        """
        params = (category.category_name, category.max_borrow_count, category.borrow_days, category.late_fee_per_day)
        return self.db.execute_non_query(sql, params)

    def update_category(self, category: ReaderCategory) -> int:
        sql = """
            UPDATE ReaderCategory
            SET CategoryName=?, MaxBorrowCount=?, BorrowDays=?, LateFeePerDay=?
            WHERE CategoryId=?
        """
        params = (category.category_name, category.max_borrow_count, category.borrow_days,
                  category.late_fee_per_day, category.category_id)
        return self.db.execute_non_query(sql, params)

    def delete_category(self, category_id: int) -> int:
        sql = "DELETE FROM ReaderCategory WHERE CategoryId=?"
        return self.db.execute_non_query(sql, (category_id,))

File: BM.DAL/ReaderCategoryService.cs
using BookManagement.DAL.Utility;
using BookManagement.Model;
using System.Data;

namespace BookManagement.DAL
{
    public class ReaderCategoryService
    {
        public DataTable GetAllCategories()
        {
            // 注意：这里假设列名是 CategoryId, CategoryName, MaxBorrowCount, BorrowDays, LateFeePerDay
            string sql = "SELECT CategoryId, CategoryName, MaxBorrowCount, BorrowDays, LateFeePerDay FROM ReaderCategory ORDER BY CategoryId";
            DataSet ds = SQLHelper.ExecuteDataSet(sql);
            return ds?.Tables?[0];
        }

        public int AddCategory(ReaderCategory category)
        {
            string sql = $@"
                INSERT INTO ReaderCategory (CategoryName, MaxBorrowCount, BorrowDays, LateFeePerDay) 
                VALUES ('{category.CategoryName}', {category.MaxBorrowCount}, {category.BorrowDays}, {category.LateFeePerDay})";
            return SQLHelper.ExecuteNonQuery(sql);
        }

        public int UpdateCategory(ReaderCategory category)
        {
            string sql = $@"
                UPDATE ReaderCategory 
                SET CategoryName='{category.CategoryName}', 
                    MaxBorrowCount={category.MaxBorrowCount}, 
                    BorrowDays={category.BorrowDays}, 
                    LateFeePerDay={category.LateFeePerDay}
                WHERE CategoryId={category.CategoryId}";
            return SQLHelper.ExecuteNonQuery(sql);
        }

        public int DeleteCategory(int categoryId)
        {
            string sql = $"DELETE FROM ReaderCategory WHERE CategoryId={categoryId}";
            return SQLHelper.ExecuteNonQuery(sql);
        }
    }
}
